feat: extract NextjsDomain from NextjsDistribution

bestickley opened the pull request:
Fixes #158

kevin-mitchell commented:
I checked out the branch, built it locally, and attempted to `cdk synth` and ended up with an error.

```Bundling warmer function...
TypeError: The "path" argument must be of type string. Received undefined
...
...cdk-nextjs-standalone/src/NextjsDistribution.ts:424:45```

which seems to refer to this line (I think?):

```
    const publicFiles = fs.readdirSync(path.join(this.props.nextjsPath, NEXTJS_BUILD_DIR, NEXTJS_STATIC_DIR), {
      withFileTypes: true,
    });
    ```

I'm guessing this is user error on my part, I'm probably doing something dumb, but I promised I'd take a look this weekend and I hit this so figured I'd throw this out there.

I'll take a closer look tomorrow (maybe I'm missing a config or something).

kevin-mitchell commented:
And of course indeed the issue was / is with the config changes required. I'm going to update my config to make sure I put the right configuration values where they need to go and will update again shortly..

kevin-mitchell commented on src/NextjsDomain.ts:
```diff
@@ -0,0 +1,145 @@
+   */
+  readonly alternateNames?: string[];
+  /**
+   * Set this option if the domain is not hosted on Amazon Route 53.
```
```suggestion
   * Set this option if the domain is not hosted on Amazon Route 53 or is hosted on Route53 but in a different account or otherwise in a way it is unaccessible. If false, the domain records will not be automatically created.
```

Maybe this is wordy, and may not totally clear, but this was something that threw me off. a bit. I think the name is still fine ("external domain"), but when I originally read this I thought "oh, no, my domain is in AWS" so set to true... but really it only matters if the account / IAM role running the CDK / CloudFormation is able to create Route53 records on behalf of this domain.

bestickley replied:
@kevin-mitchell, agree with you it's not clear. Now is the time if we're going to change the name of this variable. Do you think there is a better alternative?

bestickley replied:
One option could be to remove `isExternalDomain` and require that user runs
```ts
HostedZone.fromLookup(this, 'HostedZone', {
          domainName: this.props.domainName,
        });
```
from user-land?

bestickley replied:
@kevin-mitchell, I appreciate your thoughts. If user supplies `isExternalDomain`, then `hostedZone` isn't created, which means `Certificate` isn't created and DNS records aren't created. So it seems to me like there isn't much value to `NextjsDomain` if `isExternalDomain: true`. Do you agree?

Therefore, I think it makes sense to remove the `isExternalDomain` prop and require `hostedZone`. What do you think?

bestickley replied:
A potential issue with requiring `hostedZone`, is that if user wants/has to use another DNS provider, then how do they configure CloudFront Distribution to use that DNS? I guess they could use `defaults` to customize props passed into CloudFront Distribution? For example, if they want to provide their own `Certificate` that they created and validated themselves by adding DNS records to namecheap?

kevin-mitchell replied:
@bestickley 

I agree there isn't _much_ value in `isExternalDomain`, and at worse it can caused confusion. I'm sort of repeating what I said above but to me it seems like either somebody is sufficiently advance to be able to intuit what needs to be done for their specific use case, or setting `isExternalDomain` isn't really going to magically make everything work anyway. So from that perspective there isn't much value IMHO.

BUT, it is at least a "clue" I suppose, if somebody sets it to true they likely will also read the comment and that might give us an opportunity to explain a bit more about what the flag does, and what we'd expect them to take care of on their side if their domain really is external (e.g. manually create A/AAAA records for the CloudFront distribution, make sure there is a valid certificate, etc).

**BUT I don't think simply removing it is really an option actually**, as it would make some things impossible (this would actually break things for me)** This code would not work for me, because in my case I `isExternalDomain` as "true", BUT I pass in a certificate. If we simply removed the prop, then in my case the HostedZone.fromLookup would run `return HostedZone.fromLookup(this, 'Ho...`, which would fail. 

So I'd either keep it for now I think (it's late here in Japan so I might not be thinking perfectly, I can take another look in my morning tomorrow or of course I defer to you :))


```
  private getHostedZone(): IHostedZone | undefined {
    if (this.props.isExternalDomain) {
      return;
    } else {
      if (!this.props.hostedZone) {
        return HostedZone.fromLookup(this, 'HostedZone', {
          domainName: this.props.domainName,
        });
      } else {
        return this.props.hostedZone;
      }
    }
  }

  private getCertificate(): ICertificate | undefined {
    if (this.props.certificate) {
      return this.props.certificate;
    } else if (this.hostedZone) {
      return new Certificate(this, 'Certificate', {
        domainName: this.props.domainName,
        validation: CertificateValidation.fromDns(this.hostedZone),
      });
    } else {
      return;
    }
  }
```

bestickley replied:
@kevin-mitchell, great point. Can you send me the props you're passing with this new version?
If I'm understanding correctly, you're passing in certificate and hostedZone and having `NextjsDomain` create DNS records for you?

bestickley replied:
Because if you're not passing `hostedZone` and only passing `certificate`, then you could just pass the certificate to default CloudFront Distribution props to get the certificate to distribution and bypass `NextjsDomain`, right?

kevin-mitchell replied:
Just to add to the above in case the context is useful to anybody in the future reading this, what I do is create the certificate myself outside of the Constructs in this project, in `us-east-1` - apparently required. see: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/cnames-and-https-requirements.html but it says

> We recommend that you use a certificate issued by [AWS Certificate Manager (ACM)](https://aws.amazon.com/certificate-manager/). For information about getting a certificate from ACM, see the [AWS Certificate Manager User Guide](https://docs.aws.amazon.com/acm/latest/userguide/). To use an ACM certificate with CloudFront, **make sure you request (or import) the certificate in the US East (N. Virginia) Region (us-east-1)**.

Note that I use `CertificateValidation.fromDns()` if I am not creating a hosted zone in my account. This is the case in my "prod" accounts, because I use DNS Zone Delegation to slice off sub-domains for my other accounts (e.g. dev.domain.com), but in the case of prod my root AWS organization account owns the hosted zone so I can't (easily) add DNS records to a different account (I think there actually might be SOME way, but I'd prefer to keep things simple and for prod I don't mind the small one time setup). 

```
    // Note that we will need to manually create the CNAME records in the root / domain account, certainly
    // for "prod" because the hosted zone won't have permission to add the records in another root account.
    const certificate = new acm.Certificate(
      this,
      `${props.prefix}-domain-certificate`,
      {
        domainName: hostedZoneDomain,
        subjectAlternativeNames:
          props.domainConfig.alternateDomains.length > 0
            ? props.domainConfig.alternateDomains
            : undefined,
        validation: customDomainHostedZone
          ? CertificateValidation.fromDns(customDomainHostedZone)
          : CertificateValidation.fromDns(),
      },
    );

    new ssm.StringParameter(this, "domain-certificate-arn", {
      parameterName: `/${props.prefix}/domain/domain-certificate-arn`,
      stringValue: certificate.certificateArn,
    });
    ```

After I run the above the deployment actually stalls because I have to go and manually add the DNS records to validate the Certificate (basically add a few CNAME records into my root domain). I do this by logging into current / child AWS console and viewing the certificate in ACM, copying the values to my root account hosted zone.

Then within the actualy NextJS construct I do this:

```
const nextjs = new Nextjs(this, `${props.prefix}-web`, {
      nextjsPath: "./apps/web",
      buildCommand: "npx open-next@2.2.4 build",
      environment: {
        PARAMETERS_SECRETS_EXTENSION_LOG_LEVEL: "warn",
        ENVIRONMENT_PREFIX: props.prefix,
       ...
      },
      domainProps: {
        domainName: props.domainName,
        // this is the certificate I generated above
        certificate: domainCert,
        alternateNames: props.alternateDomainNames,
        // Setting this to true means the NextJs construct won't attempt to
        // create a hosted zone, and if we didn't configure a hosted zone
        // e.g. because prod domains are owned by root account then we'll
        // set this to true, because the hosted zone won't be findable
        // in the AWS account
        isExternalDomain: true,
      },
      defaults: {
        lambda: {
          paramsAndSecrets: paramsAndSecretsLayerVersion,
        },
        distribution: {
          domainNames: [
            "domain.io",
            "www.domain.io",
            "prod.domain.io",
          ],
        },
      },
    });

```

bestickley replied:
@kevin-mitchell, so I think with your code (thanks for sharing by the way), `NextjsDomain` is creating nothing for you. The only "value" it's adding for you is passing your certificate and domain names to the CloudFront Distribution which you could alternatively do with:

```ts
defaults: {
  distribution: {
    certificate: domainCert,
    domainNames: props.alternateDomainNames,
  },
}
```
Could you try? I know it's late for you, so no rush.

kevin-mitchell replied:
I'll look in the AM, but yes, for my _prod_ environment `NextjsDomain` does nothing for me. To be clear I do use it fully for my lower environments (because domain delegation lets my lower accounts control a slice of the domain record).

I can do what you suggest but it will fail if I don't have something set in domainProps, because it also gets the domain names from the NextjsDomain:

https://github.com/jetbridge/cdk-nextjs/blob/7772bb38ce844cc1fa57d7b8fde4a4c0dfeb1be2/src/NextjsDistribution.ts#L386

https://github.com/jetbridge/cdk-nextjs/blob/7772bb38ce844cc1fa57d7b8fde4a4c0dfeb1be2/src/NextjsDomain.ts#L52

So what you suggest might work for certificate (honestly I missed this, so thank you!) but we still rely on `NextjsDomain` to pass things through. Hopefully that makes sense.

bestickley replied:
> I can do what you suggest but it will fail if I don't have something set in domainProps, because it also gets the domain names from the NextjsDomain:

You're right, it would fail. I'm going to push a commit fixes the props interface indicating it's optional which will give me type error. Then I'll use optional chaining (`this.props.nextDomain?.domainNames`) so it doesn't break.

>  but we still rely on NextjsDomain to pass things through

I think we should change this.

bestickley replied:
I'm leaning towards removing `isExternalDomain` because it is confusing to me to have a single prop in a construct make it so that no constructs are instantiated in that construct. I could be convinced otherwise.
@revmischa, any thoughts?

kevin-mitchell replied:
> I have changed this.

FYI, maybe you didn't push yet, but I believe as of now the `nextDomain` is still where the certificate and domain names are being pulled from https://github.com/jetbridge/cdk-nextjs/pull/174/files#diff-1c29a62fdd463658d0b7511d7e6ff52f8d99a7bdb242659b103536cc5e05d63bR386-R387

bestickley replied:
@kevin-mitchell, I didn't see that `...cfDistributionProps,` was before `domainNames` and `certificate` in the object so you currently cannot override. I just fixed that.

revmischa replied:
would a prop like createRoute53Records: boolean be clearer?

bestickley replied:
@revmischa, good suggestion. `createRoute53Records` is more a explicit/procedural name. However currently I'm leaning towards removing it because I think it's a little strange/misleading/confusing to have a prop then makes it so no child constructs are created in the construct you're instantiating. Do you agree?

revmischa replied:
I'm a fan of making the typical happy path simple. Most people who deploy a nextjs site with a custom domain on AWS probably would like the route53 records created for them. They could be created in a separate construct if this one is getting too overloaded.

bestickley replied:
> would like the route53 records created for them

I completely agree! And `NextjsDomain` will do this for the user. What I'm suggesting is not to make it more difficult for the typical happy path, but more clear for the complex use case by removing `isExternalDomain` which can and has caused confusion to users. This will tell users they need to use their own constructs outside `Nextjs`

kevin-mitchell commented on src/NextjsDistribution.ts:
```diff
@@ -448,7 +392,7 @@ export class NextjsDistribution extends Construct {
       // these values can NOT be overwritten by cfDistributionProps
-      domainNames,
+      domainNames: this.props.domainNames,
       certificate: this.certificate,
```
This value is coming from an uninitiated instance variable (as far as I can tell!).

`this.certificate,` needs to be created or passed in from the Domain construct.

kevin-mitchell reviewed (changes requested):
I ran a diff between a project I had deployed a few versions back and this PR, this is the only diff I saw related to the Distribution...

```
[~] AWS::CloudFront::Distribution prod-web/Distribution/Distribution prodwebDistribution1CB9742B 
 └─ [~] DistributionConfig
     └─ [-] Removed: .ViewerCertificate
```

At first I thought maybe this was fine... but no, this is an issue. Basically the certificate needs to be supplied to the Distribution, because if the distribution doesnt' have the certificate set here then the resulting Distribution will look like this:

<img width="731" alt="Screenshot 2023-11-20 at 5 34 56 PM" src="https://github.com/jetbridge/cdk-nextjs/assets/21270844/f6611bdb-9b0a-4ea5-b283-d4f01645b643">

Resulting in...
<img width="763" alt="Screenshot 2023-11-20 at 5 34 43 PM" src="https://github.com/jetbridge/cdk-nextjs/assets/21270844/2c203bb7-feaf-4970-9857-78362361c301">


Repeating myself here I think but just adding the prop to the NextjsDistribution and then passing the value from the new NextjsDomain, e.g. adding `certificate: this.domain?.certificate,` below I think would do the trick?

```
...
 if (this.props.domainProps) {
      this.domain = new NextjsDomain(this, 'Domain', this.props.domainProps);
    }
    this.distribution = new NextjsDistribution(this, 'Distribution', {
      basePath: props.basePath,
      distribution: props.distribution,
      ...props.defaults?.distribution,
      staticAssetsBucket: this.staticAssets.bucket,
      nextBuild: this.nextBuild,
      serverFunction: this.serverFunction.lambdaFunction,
      imageOptFunction: this.imageOptimizationFunction,
      domainNames: this.domain?.domainNames,
      certificate: this.domain?.certificate,
    });
    if (this.domain) {
      this.domain.createDnsRecords(this.distribution.distribution);
    }
...
    ```

bestickley commented:
> I'm guessing this is user error on my part, I'm probably doing something dumb, but I promised I'd take a look this weekend and I hit this so figured I'd throw this out there.
I think this was an issue on my end! I'm not passing `nextjsPath` into `NextjsDistribution`

kevin-mitchell commented:
> I think this was an issue on my end! I'm not passing `nextjsPath` into `NextjsDistribution`

that is true, I assumed that was intentional on your part. If not, then yeah that's missing!

bestickley commented on src/NextjsDistribution.ts:
```diff
@@ -61,31 +56,14 @@ export interface NextjsDistributionProps {
    */
   readonly distribution?: NextjsProps['distribution'];
+  /**
+   * Alternative domain names for CloudFront Distribution
```
I've removed `domainNames` prop and am now just accepting `nextDomain` which has `domainNames` on it.

bestickley commented:
@kevin-mitchell, go ahead and try out this new version and let me know what you think

kevin-mitchell commented:
> @kevin-mitchell, go ahead and try out this new version and let me know what you think

Just redeployed, my certificate was added back (yay!) and things seem to work for me!

Thanks a ton for your work, sorry again I didn't end up putting up a PR for this! FYI I'm hoping to get back to paying a bit more attention to this in the next few weeks, just got a bit busy.

kevin-mitchell reviewed (approved):
I was able to deploy and domain configs are working for me with the new structure / changes!

bestickley commented:
> Thanks a ton for your work, sorry again I didn't end up putting up a PR for this! FYI I'm hoping to get back to paying a bit more attention to this in the next few weeks, just got a bit busy.

No problem at all. Totally understand that open source contributions ebb and flow :)

bestickley commented:
@kevin-mitchell and @revmischa, I noticed that the `domainAliases` property creates an L3 [HttpsRedirect](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_route53_patterns-readme.html#https-redirect) construct which from the docs:
> The HttpsRedirect constructs creates:
- Amazon CloudFront distribution - makes website available from data centres around the world
- Amazon S3 bucket - empty bucket used for website hosting redirect (websiteRedirect) capabilities.
- Amazon Route 53 A/AAAA Alias records - routes traffic to the CloudFront distribution
- AWS Certificate Manager certificate - SSL/TLS certificate used by CloudFront for your domain

I don't think users passing in a value to `domainAliases` will expect all of those resources to be created from one prop, do you? What's the use case for this? The comment says, "An alternative domain to be assigned to the website URL. Visitors to the alias will be redirected to the main domain. (ie. `www.domain.com`). Use this to create a `www.` version of your domain and redirect visitors to the root domain.", but if I want trying to create a `www.` version of my website, I don't think I'd want a separate distribution created.

What do you guys think?

bestickley commented:
Some other changes:
- I've improved the TS Doc comment to be more clear/detailed (IMO) so please review.
- Currently, the certificate that is created isn't a wildcard certificate. This means that `alternateNames` cannot be used as explained [here](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/CNAMEs.html#alternate-domain-names-requirements). (Has anyone used this feature before???) Therefore I recommend we create a wildcard certificate which simply means prefixing the domain with a wildcard like `*.example.com`. `isWildcardCertificate` will control this behavior and by default will be `true`
- `createDnsRecords` was only creating DNS records for the `domainName` passed in, not the `alternativeNames`. I'd expect as a user of this construct that DNS records would be created for any domains passed into `NextjsDomain` so I'm now creating A and AAAA records for those alternative domain names.

These are more changes than I wanted to make, but I want all props of this construct to be functional, make sense, and offer no surprises.

@kevin-mitchell, do you mind re-testing?

kevin-mitchell commented:
@bestickley testing now. Actually I just ran a diff and I see:

```
[~] AWS::CloudFront::Distribution prod-web/Distribution/Distribution prodwebDistribution1CB9742B 
 └─ [~] DistributionConfig
     ├─ [-] Removed: .Aliases
     └─ [-] Removed: .ViewerCertificate
```

but I think I'm perhaps not entirely certain what the expected config is at this point for the use case I described previously.  The above result is from this config:

```
const nextjs = new Nextjs(this, `${props.prefix}-web`, {
      nextjsPath: "./apps/web",
      environment: {
        PARAMETERS_SECRETS_EXTENSION_LOG_LEVEL: "warn",
       ...
      },
      defaults: {
        lambda: {
          paramsAndSecrets: paramsAndSecretsLayerVersion,
        },
        distribution: {
          certificate: domainCert,
          domainNames: [
            "domain.io",
            "www.domain.io",
            "prod.domain.io",
          ],
        },
      },
    });
```

Note I did not include the `domainProps`. This is expected I think looking at the code, but just confirming.

If I _do_ include `domainProps`, e.g. 

```
const nextjs = new Nextjs(this, `${props.prefix}-web`, {
     ...snip snip...
      domainProps: {
        certificate: domainCert,
        isWildcardCertificate: true,
        domainName: "domain.io",
        alternateNames: ["www.domain.io", "prod.domain.io"],
      },
      defaults: {
        lambda: {
          paramsAndSecrets: paramsAndSecretsLayerVersion,
        },
        distribution: {
          certificate: domainCert,
          domainNames: [
            "domain.io",
            "www.domain.io",
            "prod.domain.io",
          ],
        },
      },
    });
```

Then this fails with

```
[Error at /Web-prod/prod-web/Domain] Found zones: [] for dns:domain.io, privateZone:undefined, vpcId:undefined, but wanted exactly 1 zone
```

I haven't stepped through the code yet but can dig deeper tomorrow. **I'm guessing this is user error**.

----

One small comment re:

> but if I want trying to create a www. version of my website, I don't think I'd want a separate distribution created.

I'm fairly certain that's what the [AWS construct does](https://github.com/aws/aws-cdk/blob/v2.110.0/packages/aws-cdk-lib/aws-route53-patterns/lib/website-redirect.ts#L77), but I think you're saying just remove the feature entirely, if somebody wants a redirect they can just add it themselves easily with this construct? That makes sense to me, I'm all for less magic.

kevin-mitchell commented:
@bestickley just to follow up from my comment yesterday, but at least for my use case (and going back to what I thought you were originally intending), can you just add an OPtiONAL override default property to the NextjsDistribution props? If somebody supplies these values then they'll be used instead of the NextjsDomain derived values. 

In my situation at least this allows me to completely skip over creating the NextjsDomain for my production environment by simply not supplying `domainProps` to the Nextjs construct.

(is there a better way of supplying a patch without opening a PR against your PR or something? I should probably know this :/)


```
diff --git a/src/NextjsDistribution.ts b/src/NextjsDistribution.ts
index 28dfbfb..b73f76f 100644
--- a/src/NextjsDistribution.ts
+++ b/src/NextjsDistribution.ts
@@ -1,6 +1,7 @@
 import * as fs from 'node:fs';
 import * as path from 'path';
 import { Duration, Fn, RemovalPolicy } from 'aws-cdk-lib';
+import { ICertificate } from 'aws-cdk-lib/aws-certificatemanager';
 import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
 import { Distribution, ResponseHeadersPolicy } from 'aws-cdk-lib/aws-cloudfront';
 import * as origins from 'aws-cdk-lib/aws-cloudfront-origins';
@@ -95,6 +96,16 @@ export interface NextjsDistributionProps {
    * Must be provided if you want to serve static files.
    */
   readonly staticAssetsBucket: s3.IBucket;
+  /**
+   * Explicitly set the domain names to be associated with the Distribution. Note that
+   * if this is set it will override anything created by {@see NextjsDomain }, but this may
+   * be what you want if you are managing the domains outside of the {@see Nextjs } construct.
+   */
+  readonly domainNames: string[];
+  /**
+   * Explicitly supply the certificate to be assigned to the Distribution.
+   */
+  readonly certificate: ICertificate;
 }
 
 /**
@@ -378,8 +389,8 @@ export class NextjsDistribution extends Construct {
       // defaultRootObject: "index.html",
       defaultRootObject: '',
       minimumProtocolVersion: cloudfront.SecurityPolicyProtocol.TLS_V1_2_2021,
-      domainNames: this.props.nextDomain?.domainNames,
-      certificate: this.props.nextDomain?.certificate,
+      domainNames: this.props.domainNames || this.props.nextDomain?.domainNames,
+      certificate: this.props.certificate || this.props.nextDomain?.certificate,
 
       // Override props.
       ...cfDistributionProps,

```

bestickley commented:
Thanks for reporting the error and your suggestion. I'm on travel the next couple days. I will respond early next week.

bestickley commented:
Also, because `NextjsDomain` is not created by default, what do you think about just leaving it as a separate construct that a user can instantiate if needed? This makes more sense to me than requiring the user to pass in props to top level `Nextjs` construct.

bestickley commented:
re: the diff you saw above:
```bash
[~] AWS::CloudFront::Distribution prod-web/Distribution/Distribution prodwebDistribution1CB9742B 
 └─ [~] DistributionConfig
     ├─ [-] Removed: .Aliases
     └─ [-] Removed: .ViewerCertificate
```
It seems like the `defaults.distribution` are not being passed to the `Distribution` construct. Can you look at outputted CloudFormation JSON template to confirm? Will be in cdk.out/ folder.

bestickley commented:
I added the below TS Doc to `NextjsDomain`:
```ts
/**
 * Use a custom domain with `Nextjs`. Requires a Route53 hosted zone to have been
 * created within the same AWS account. For DNS setups where you cannot use a
 * Route53 hosted zone in the same AWS account, use the `defaults.distribution`
 * prop of {@link NextjsProps}.
 *
 * See props interface TS Doc comments for detailed docs on how to customize.
 * This construct is helpful to user to not have to worry about interdependencies
 * between Route53 Hosted Zone, CloudFront Distribution, and Route53 Hosted Zone Records.
 */
export class NextjsDomain extends Construct {
```
and this to `NextjsProps`:
```ts
/**
   * Props to configure {@link NextjsDomain}. See details on how to customize at
   * {@link NextjsDomainProps}
   */
  readonly domainProps?: NextjsDomainProps;
```
Am I missing anything?

bestickley commented:
@kevin-mitchell, please test out with new props as I explain above and let me know if it works for you. I think this is ready.

kevin-mitchell commented:
@bestickley first thanks as always for the work on the project.

Second, you changes worked both:

1. In my `prod` use case where I'm creating everything and just passing the values to the Distribution construct to have the certificate and domain names set in CloudFront. This is working perfectly now. Thank you!
2. In my `dev` use case, where the hosted zone is looked up and Route53 records are added to point to the new Distribution. Awesome!

RE: `defaults`, I have a similar question / general lack of love for the "defaults" / deeply nested values. It feels weird to me (I'll admit I haven't thought very hard about it) to have `domainProps` as a top level, then `defaults` elsewhere. But "overrides" seems like that'll address that!

Otherwise the only annoying that that I'm still not sure about is the fact that `isWildcardCertificate` is required even when you provide a certificate. I guess this is just a matter of it being complicated to support different use cases, but my issue here is that it's weird to have to pass in `isWildcardCertificate` when you're passing in a certificate that you already created. If you looked at where `isWildcardCertificate` is used, it's only used in the `get` `certificateDomainName`, which ultimately is only used if don't explicitly pass in a certificate:

```
...
if (!this.props.certificate) {
      return new Certificate(this, 'Certificate', {
        domainName: this.certificateDomainName,
        validation: CertificateValidation.fromDns(this.hostedZone),
      });
    }
...
  ```

But perhaps this is a can that could be kicked down the road? 

Comments look great and are much improved IMHO.

Thanks again!

👍

bestickley commented:
@kevin-mitchell, so glad they both worked!

> I have a similar question / general lack of love for the "defaults" / deeply nested values. It feels weird to me (I'll admit I haven't thought very hard about it) to have domainProps as a top level, then defaults elsewhere. But "overrides" seems like that'll address that!

So `domainProps` I think should be separate from `defaults` or the future `overrides` because `domainProps` controls whether a secondary construct/feature of `Nextjs` is created. `defaults` customizes how primary constructs/features are created. We could put `domainProps` in `overrides` in future but semantically it wouldn't make sense because you're not overriding anything, you're just creating something. Make sense?

> Otherwise the only annoying that that I'm still not sure about is the fact that isWildcardCertificate is required even when you provide a certificate.

I don't think this is true. If you pass in your own certificate, that prop should have no impact because `isWildcardCertificate` is used in the getter `certificateDomainName` which is only used in the construction of `Certificate` if `certificate` isn't passed in. Right?

bestickley commented:
Thinking through `isWildcardCertificate` more, I don't think it's wise to have a wildcard certificate created by default. It can be a security issue to some and this construct should take a "secure by default" approach when possible. I've replaced it with `certificateDomainName` so that user can easily specify `domainName` when `Certificate` is created. I added a note on `alternateNames` that `certificateDomainName` will need to be specified for `alternateNames` to work. Make sense?

kevin-mitchell commented:
> ... If you pass in your own certificate, that prop should have no impact ...

Sorry possible miscommunication on my part, but this is exactly what I'm saying. It has no impact, but is required. It's a really minor thing, just feels a bit weird to me to have a required property that isn't actually required. _That said your comment on the property is clear_, so I feel like it's not a problem.

I think in my mind ideally that property would only be required in a situation where the other properties passed in would be such that a certificate WOULD be created. e.g. the type to `domainProps.certificate` could be `ICertificate | CertificateCreationOptions`, where `CertificateCreationOptions` could be `{ isWildcardCertificate: boolean }` - this might actually be a terrible example but hopefully the idea at least makes sense :)

Again though, I feel your comment makes it clear enough that the property isn't used if you pass in a certificate, and I'd guess your trigger finger on the merge button is getting pretty itchy by this point on this PR :)

kevin-mitchell commented:
@bestickley I missed your follow up comment (sorry!), which makes my last comment irrelevant. 🚢 :shipit:  :)
